deploy/buildtools/buildconfig.py

```diff
@@ -1,177 +1,151 @@
-""" This converts the build configuration files into something usable by the
-manager """
-
 from time import strftime, gmtime
-import configparser
 import pprint
+from importlib import import_module
 
-from runtools.runtime_config import RuntimeHWDB
 from awstools.awstools import *
+from buildtools.buildfarmhostdispatcher import BuildFarmHostDispatcher
+
+# typing imports
+from typing import Set, Type, Any, Optional, Dict, TYPE_CHECKING
+# needed to avoid type-hint circular dependencies
+# TODO: Solved in 3.7.+ by "from __future__ import annotations" (see https://stackoverflow.com/questions/33837918/type-hints-solve-circular-dependency)
+#       and normal "import <module> as ..." syntax (see https://www.reddit.com/r/Python/comments/cug90e/how_to_not_create_circular_dependencies_when/)
+if TYPE_CHECKING:
+    from buildtools.buildconfigfile import BuildConfigFile
+else:
+    BuildConfigFile = object
+
+def inheritors(klass: Type[Any]) -> Set[Type[Any]]:
+    """Determine the subclasses that inherit from the input class.
+
+    Args:
+        klass: Input class.
+
+    Returns:
+        Set of subclasses that inherit from input class.
+    """
+    subclasses = set()
+    work = [klass]
+    while work:
+        parent = work.pop()
+        for child in parent.__subclasses__():
+            if child not in subclasses:
+                subclasses.add(child)
+                work.append(child)
+    return subclasses
 
 class BuildConfig:
-    """ This represents a SINGLE build configuration. """
-    def __init__(self, name, buildconfigdict, launch_time):
+    """Represents a single build configuration used to build RTL, drivers, and bitstreams.
+
+    Attributes:
+        name: Name of config i.e. name of `config_build_recipe.yaml` section.
+        build_config_file: Pointer to global build config file.
+        TARGET_PROJECT: Target project to build.
+        DESIGN: Design to build.
+        TARGET_CONFIG: Target config to build.
+        deploytriplet: Deploy triplet override.
+        launch_time: Launch time of the manager.
+        PLATFORM_CONFIG: Platform config to build.
+        s3_bucketname: S3 bucketname for AFI builds.
+        post_build_hook: Post build hook script.
+        build_farm_host: Name of build farm host type.
+        build_farm_host_dispatcher: Build farm host dispatcher object.
+    """
+    name: str
+    build_config_file: BuildConfigFile
+    TARGET_PROJECT: Optional[str]
+    DESIGN: str
+    TARGET_CONFIG: str
+    deploytriplet: Optional[str]
+    launch_time: str
+    PLATFORM_CONFIG: str
+    s3_bucketname: str
+    post_build_hook: str
+    build_farm_host: str
+    build_farm_host_dispatcher: BuildFarmHostDispatcher
+
+    def __init__(self, name: str, recipe_config_dict: Dict[str, Any], build_farm_hosts_config_file: Dict[str, Any], build_config_file: BuildConfigFile, launch_time: str) -> None:
```

Comment:
nit: split this line into multiple.  I can't read it without scrolling in GH.

we could (in a separate, future PR) start using black (or something else) to autoformat our python.